### PR TITLE
chore: consolidate download service

### LINKED ISSUE #1242: feat: Model Pull has clear API and CLI to support Huggingface Repos
## Goal 

- `cortex model pull` should have clear APIs that support different model repo sources
- e.g. Huggingface, Cortex Hub

## Tasklist

- [x] #1393 
- [x] #1394 
- [x] #1395
- [x] #1398

## CLI
```
# Pulls immediately
cortex model pull <huggingface_url_sppecific_gguf>

# Lets user select quantization using CLI
cortex model pull <huggingface_url> 

# NOT SURE: Do we need an "info" equivalent?
# Gets repo type (e.g. GGUF, in future ONNX, TensorRT-LLM, dumps possible versions)
# Will power the "select quantization"
cortex model info <huggingface_url>
cortex model info <cortex_repo_url>    # Dumps tags
```

## API
- Is there any change necessary to the models pull API?
- https://cortex.so/api-reference/#tag/models/post/v1/models/%7BmodelId%7D/pull
```
- How do we represent Huggingface strings?
- How do we handle Huggingface Repos (i.e. where user has to select quantization)? 
```

## Key Questions

- Does Cortex need an API to get Huggingface Repo metadata?
- Does this need to be bubbled to Jan, to support the Huggingface Repo page?

## Linked Issues

- https://github.com/janhq/cortex.cpp/issues/1006
- https://github.com/janhq/cortex.cpp/issues/1145
- https://github.com/janhq/cortex.cpp/issues/1017
- And your PR: https://github.com/janhq/cortex.cpp/pull/1239

## Jan's Requirements

1. User enters Huggingface URL in import box
2. User clicks deep link from Huggingface

Cortex should support an API, that can support the following UI: 

Image

### LINKED ISSUE #1393: idea: [DownloadEvent] Adding websocket to publish download event to client
### Problem Statement

- Currently we don't send any event to client. So, client can't update UI accordingly.
- Drogon does not support SSE.

### Feature Idea

- Publishing events via websocket at `ws://127.0.0.1:3928/events`

### LINKED ISSUE #1394: idea: [DownloadEvent] Adding event queue to process event asynchronously
### Problem Statement

If we write a simple callback, the thread of event emitter will have to wait for the event listener to finished consuming events. This might cause hang.

### Feature Idea

Using [eventpp package](https://vcpkg.link/ports/eventpp) to have an `EventQueue` which acts as a buffer in between `EventEmitter` and `EventListener`

### LINKED ISSUE #1395: idea: [DownloadEvent] Wiring up event library and download service
### Problem Statement

Wiring up event library and download service

### Feature Idea

- Define reusable events (DownloadEvent, ModelEvent, HardwareEvent, etc.)
- Define an easy API to publish those events
- Broadcast events via websocket

### PR DESCRIPTION
**Pull request by namchuai**
## Describe Your Changes

- Consolidate `Download service` for sending event to client.
- Clean up warning while building.
- Adding websocket
- Adding event queue
- Update Download service

## Screenshots
Download events via websocket
Screenshot 2024-10-03 at 09 10 39

## Fixes Issues

- fixes #1242 
- fixes #1393 
- fixes #1394 
- fixes #1395 

## Self Checklist

- [ ] Added relevant comments, esp in complex areas
- [ ] Updated docs (for bug fixes / features)
- [ ] Created issues for follow-up changes or refactoring needed

### CODE REVIEW

File: engine/controllers/events.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  void handleNewConnection(const HttpRequestPtr& req,
+                           const WebSocketConnectionPtr& wsConnPtr) override;
+
+  void handleConnectionClosed(const WebSocketConnectionPtr& wsConnPtr) override;
+
+ private:
+  void broadcast(const std::string& message);
+
+  std::shared_ptr<EventQueue> event_queue_;
+  std::set<WebSocketConnectionPtr> connections_;
```
Comment (vansangpfiev) [inline engine/controllers/events.h]: Just curious, why do we need to use std::set?
If we have to use `set` data structure, I think can use std::unordered_set

File: engine/services/download_service.h [hunk 2]
```diff
@@ -77,9 +43,55 @@ class DownloadService {
...
-  cpp::result<bool, std::string> Download(
-      const std::string& download_id, const DownloadItem& download_item,
-      bool allow_resume) noexcept;
+  cpp::result<bool, std::string> Download(const std::string& download_id,
+                                          const DownloadItem& download_item,
+                                          bool allow_resume) noexcept;
 
   curl_off_t GetLocalFileSize(const std::filesystem::path& path) const;
+
+  std::optional<std::shared_ptr<EventQueue>> event_queue_;
```
Comment (vansangpfiev) [inline engine/services/download_service.h]: I think no need to use `std::optional` here, we can check nullptr

File: engine/services/download_service.h [hunk 3]
```diff
@@ -77,9 +43,55 @@ class DownloadService {
...
+  std::optional<std::string> active_download_item_id_;
+
+  /**
+   * Invoked when download is completed (both failed or success)
+   */
+  void CleanUp(const std::string& task_id);
+
+  static int ProgressCallback(void* ptr, double dltotal, double dlnow,
+                              double ultotal, double ulnow) {
+    DownloadService* service = static_cast<DownloadService*>(ptr);
```
Comment (vansangpfiev) [inline engine/services/download_service.h]: Can use `auto`?

File: engine/utils/event_processor.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+using Event = cortex::event::Event;
+using DownloadEvent = cortex::event::DownloadEvent;
+using EventQueue = eventpp::EventQueue<std::string, void(DownloadEvent)>;
+
+class EventProcessor {
+ public:
+  EventProcessor(std::shared_ptr<EventQueue> queue)
+      : event_queue_(std::move(queue)), running_(true) {
+    thread_ = std::thread([this]() {
+      while (running_) {
```
Comment (vansangpfiev) [inline engine/utils/event_processor.h:21]: Do we have a way to use `std::condition_variable` in this case? I think can check for `!emptyQueue()`

File: engine/controllers/events.h (hunk 1 above)
Comment (namchuai) [reply engine/controllers/events.h]: Good catch! I will change to unordered_set

File: engine/utils/event_processor.h (hunk 4 above)
Comment (namchuai) [reply engine/utils/event_processor.h:21]: If we make using `!emptyQueue()`, then when the queue is empty, the thread will be dismissed

Comment (namchuai) [reply engine/utils/event_processor.h:21]: Update remove sleep

**Review by vansangpfiev — APPROVED** [review]

**Pull request merged**